### PR TITLE
Issue #236

### LINKED ISSUE #236: Crash in MergeFiles benchmark
```
BenchmarkMergeFiles/one_repeated_level
panic: runtime error: index out of range [45985] with length 40564

goroutine 304 [running]:
github.com/segmentio/parquet-go.(*byteArrayColumnBuffer).Less(0x4300?, 0x0?, 0x5?)
	/home/achille_roussel/go/src/github.com/segmentio/parquet-go/column_buffer.go:1494 +0x17f
github.com/segmentio/parquet-go.nullsGoLast({0xa4cc30?, 0xc0000743c0?}, 0x0?, 0x180?, 0x0?, 0x5f?, 0xde?)
	/home/achille_roussel/go/src/github.com/segmentio/parquet-go/column_buffer.go:121 +0x42
github.com/segmentio/parquet-go.(*repeatedColumnBuffer).Less(0xc0000fc120, 0x72b618?, 0x3d6f)
	/home/achille_roussel/go/src/github.com/segmentio/parquet-go/column_buffer.go:597 +0x1a6
github.com/segmentio/parquet-go.(*Buffer).Less(0x1?, 0x1?, 0x1?)
	/home/achille_roussel/go/src/github.com/segmentio/parquet-go/buffer.go:158 +0x83
sort.medianOfThree({0xa3f670, 0xc0000fc000}, 0xc0009fa180?, 0xc0?, 0xc0000bfd30?)
	/usr/local/go/src/sort/sort.go:89 +0x42
sort.doPivot({0xa3f670, 0xc0000fc000}, 0x0, 0x4636)
	/usr/local/go/src/sort/sort.go:116 +0xd8
sort.quickSort({0xa3f670, 0xc0000fc000}, 0x1?, 0x1?, 0xc0000bfdb0?)
	/usr/local/go/src/sort/sort.go:203 +0x6c
sort.Sort({0xa3f670, 0xc0000fc000})
	/usr/local/go/src/sort/sort.go:231 +0x53
github.com/segmentio/parquet-go_test.BenchmarkMergeFiles.func1(0xc0000c5200)
	/home/achille_roussel/go/src/github.com/segmentio/parquet-go/merge_test.go:93 +0x2ff
testing.(*B).runN(0xc0000c5200, 0x1)
	/usr/local/go/src/testing/benchmark.go:193 +0x102
testing.(*B).run1.func1()
	/usr/local/go/src/testing/benchmark.go:233 +0x59
created by testing.(*B).run1
	/usr/local/go/src/testing/benchmark.go:226 +0x9c
exit status 2
```

### PR DESCRIPTION
**Pull request by achille-roussel**
Fixes #236 

I tracked down the bugs that were causing crashes in tests for merging files containing repeated columns, now all these are passing:
```
goos: linux
goarch: amd64
pkg: github.com/segmentio/parquet-go
cpu: Intel(R) Xeon(R) CPU @ 2.60GHz
BenchmarkMergeFiles
BenchmarkMergeFiles/one_repeated_level
BenchmarkMergeFiles/one_repeated_level/groups=1,rows=10000
BenchmarkMergeFiles/one_repeated_level/groups=1,rows=10000         	   13525	     86840 ns/op	  11515444 row/s
BenchmarkMergeFiles/one_repeated_level/groups=2,rows=20000
BenchmarkMergeFiles/one_repeated_level/groups=2,rows=20000         	   13884	     87659 ns/op	  11407823 row/s
BenchmarkMergeFiles/one_repeated_level/groups=3,rows=30000
BenchmarkMergeFiles/one_repeated_level/groups=3,rows=30000         	   13602	     87728 ns/op	  11398886 row/s
BenchmarkMergeFiles/two_repeated_levels
BenchmarkMergeFiles/two_repeated_levels/groups=1,rows=10000
BenchmarkMergeFiles/two_repeated_levels/groups=1,rows=10000        	    2251	    516019 ns/op	   1937920 row/s
BenchmarkMergeFiles/two_repeated_levels/groups=2,rows=20000
BenchmarkMergeFiles/two_repeated_levels/groups=2,rows=20000        	    2176	    521545 ns/op	   1917388 row/s
BenchmarkMergeFiles/two_repeated_levels/groups=3,rows=30000
BenchmarkMergeFiles/two_repeated_levels/groups=3,rows=30000        	    2138	    522151 ns/op	   1915163 row/s
BenchmarkMergeFiles/three_repeated_levels
BenchmarkMergeFiles/three_repeated_levels/groups=1,rows=10000
BenchmarkMergeFiles/three_repeated_levels/groups=1,rows=10000      	    2168	    514332 ns/op	   1944277 row/s
BenchmarkMergeFiles/three_repeated_levels/groups=2,rows=20000
BenchmarkMergeFiles/three_repeated_levels/groups=2,rows=20000      	    2197	    525911 ns/op	   1901471 row/s
BenchmarkMergeFiles/three_repeated_levels/groups=3,rows=30000
BenchmarkMergeFiles/three_repeated_levels/groups=3,rows=30000      	    2247	    530843 ns/op	   1883804 row/s
BenchmarkMergeFiles/repeated_key-value_pairs
BenchmarkMergeFiles/repeated_key-value_pairs/groups=1,rows=10000
BenchmarkMergeFiles/repeated_key-value_pairs/groups=1,rows=10000   	    1339	    865039 ns/op	   1156023 row/s
BenchmarkMergeFiles/repeated_key-value_pairs/groups=2,rows=20000
BenchmarkMergeFiles/repeated_key-value_pairs/groups=2,rows=20000   	    1303	    858523 ns/op	   1164796 row/s
BenchmarkMergeFiles/repeated_key-value_pairs/groups=3,rows=30000
BenchmarkMergeFiles/repeated_key-value_pairs/groups=3,rows=30000   	    1312	    865284 ns/op	   1155695 row/s
BenchmarkMergeFiles/map_of_repeated_values
BenchmarkMergeFiles/map_of_repeated_values/groups=1,rows=10000
BenchmarkMergeFiles/map_of_repeated_values/groups=1,rows=10000     	    2316	    486068 ns/op	   2057332 row/s
BenchmarkMergeFiles/map_of_repeated_values/groups=2,rows=20000
BenchmarkMergeFiles/map_of_repeated_values/groups=2,rows=20000     	    2264	    498432 ns/op	   2006300 row/s
BenchmarkMergeFiles/map_of_repeated_values/groups=3,rows=30000
BenchmarkMergeFiles/map_of_repeated_values/groups=3,rows=30000     	    2342	    487309 ns/op	   2052094 row/s
PASS
ok  	github.com/segmentio/parquet-go	23.299s
```

### CODE REVIEW

**Comment by kevinburkesegment** [comment]
Huh. Should we enable those tests as part of automated CI now?

File: column_buffer.go [hunk 1]
```diff
@@ -590,10 +590,10 @@ func (col *repeatedColumnBuffer) Less(i, j int) bool {
...
 
 	for k := 0; k < row1Length && k < row2Length; k++ {
-		x := int(row1.offset) + k
-		y := int(row2.offset) + k
-		definitionLevel1 := col.definitionLevels[j+k]
-		definitionLevel2 := col.definitionLevels[j+k]
+		x := int(row1.baseOffset)
+		y := int(row2.baseOffset)
+		definitionLevel1 := col.definitionLevels[int(row1.offset)+k]
+		definitionLevel2 := col.definitionLevels[int(row2.offset)+k]
```
Comment (achille-roussel) [inline column_buffer.go:596]: The wrong offsets were being used: `.offset` is the position within the definition levels, `.baseOffset` is the position of the row values in the underlying column buffer.

File: column_buffer.go [hunk 2]
```diff
@@ -617,49 +617,40 @@ func (col *repeatedColumnBuffer) Swap(i, j int) {
...
-		numValues += len(row)
-		if len(col.buffer) > maxNumValues {
-			maxNumValues = len(col.buffer)
+
+		if len(col.buffer) > maxRowLen {
+			maxRowLen = len(col.buffer)
 		}
+
+		numValues += j - i
+		i = j
```
Comment (achille-roussel) [inline column_buffer.go:645]: There were issues with the way continuations of rows were handled (e.g. write values of a row to a column buffer in multiple calls to `WriteValues`).

**Comment by achille-roussel** [comment]
> Huh. Should we enable those tests as part of automated CI now?

I couldn't come up with a tests that reproduces the issue, only the benchmarks were. Maybe we can run them with `-benchtime 1x` so they don't take hours to complete?

**Pull request merged**